apps/condo/domains/user/utils/serverSchema/auth.js

```diff
@@ -0,0 +1,410 @@
+const { get } = require('lodash')
+
+const conf = require('@open-condo/config')
+const { itemsQuery, getSchemaCtx } = require('@open-condo/keystone/schema')
+
+const { normalizePhone } = require('@condo/domains/common/utils/phone')
+const { AUTH_COUNTER_LIMIT_TYPE } = require('@condo/domains/user/constants/limits')
+const { ConfirmPhoneAction } = require('@condo/domains/user/utils/serverSchema')
+const { RedisGuard } = require('@condo/domains/user/utils/serverSchema/guards')
+const { generateSimulatedToken } = require('@condo/domains/user/utils/tokens')
+
+
+const ERROR_TYPES = {
+    NOT_ENOUGH_AUTH_FACTORS: 'NOT_ENOUGH_AUTH_FACTORS',
+}
+
+const GUARD_DEFAULT_WINDOW_SIZE_IN_SEC = 60 * 60 // 1 hour in sec
+const GUARD_DEFAULT_WINDOW_LIMIT = 10
+
+/**
+ * @typedef {Object} GuardQuota
+ * @property {number} windowSizeInSec The window size in seconds
+ * @property {number} windowLimit Attempts limit during the window
+ */
+
+/**
+ * @type {{ ip?: Record<string, GuardQuota>, user?: Record<string, GuardQuota>, phone?: Record<string, GuardQuota>, email?: Record<string, GuardQuota>, default?: GuardQuota }}
+ *
+ * Examples:
+ *
+ * 1.1 Change only window size for ip
+ * { "ip": { "*.*.*.*": { windowSizeInSec: 3600 } } }
+ *
+ * 1.2 Change only limit for ip
+ * { "ip": { "*.*.*.*": { windowLimit: 60 } } }
+ *
+ * 1.2 Change window size and limit for ip
+ * { "ip": { "*.*.*.*": { windowSizeInSec: 3600, windowLimit: 60 } } }
+ *
+ * 2 Change window size and limit for user id
+ * { "user": { "********-****-****-****-************": { windowSizeInSec: 3600, windowLimit: 60 } } }
+ *
+ * 3 Change window size and limit for user phone
+ * { "phone": { "+7**********": { windowSizeInSec: 3600, windowLimit: 60 } } }
+ *
+ * 4 Change window size and limit for user email
+ * { "email": { "example@mail.com": { windowSizeInSec: 3600, windowLimit: 60 } } }
+ *
+ * 4 Change window size and limit for user email and phone
+ * { "phone": { "+7**********": { windowSizeInSec: 3600, windowLimit: 60 } }, "email": { "example@mail.com": { windowSizeInSec: 3600, windowLimit: 60 } } }
+ */
+let customQuotas
+try {
+    customQuotas = JSON.parse(conf.AUTH_GUARD_CUSTOM_QUOTAS)
+} catch (e) {
+    customQuotas = {}
+}
+
+
+const redisGuard = new RedisGuard()
+
+
+/**
+ * The following guards are activated:
+ * 1. Guard by ip (each request)
+ * 2. Guard by authed user id (if user authed)
+ * 3. Guard by phone (if phone passed)
+ * 4. Guard by email (if email passed)
+ *
+ * @param {{ phone?: string, email?: string, userType: 'staff' | 'resident' | 'service' }} userIdentity
+ * @param context
+ * @return {Promise<void>}
+ */
+async function authGuards (userIdentity, context ) {
+    if (!context) throw new Error('context cannot be empty')
```

Comment:
why not throw error if userIdentity is null? This way error will be more explicit, then one thrown in line 82

Reply:
I don't quite understand what you're talking about. Right now, it's happening just like you wrote.
There is even a test that checks this behavior)